### PR TITLE
Use recommended configurations to test plugin

### PR DESCRIPTION
**Pull request by darxriggs**
This pull request contains all changes required for using `buildPlugin(configurations: buildPlugin.recommendedConfigurations())`.

The first relevant commit is "Use @DataBoundConstructor for every AuthorizeProjectStrategy in tests". As of 2019-08-27 there are 5 other commits contained because the target branch is not yet rebased on the _master_ branch.

### CODE REVIEW

**Comment by darxriggs** [comment]
@ikedam on [build #1](https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fauthorize-project-plugin/detail/PR-40/1/pipeline) you can see that using `buildPlugin(configurations: buildPlugin.recommendedConfigurations())` now runs fine because of all the changes applied.

One way to get this onto _master_ would be that I put commit e61a835d1a3eba702333421640f2c8949008fa90 that changes the Jenkinsfile on top and change the target branch to _master_. This would a) keep the history clean and b) make tests for every commit in this pull request run successfully (useful in case `git bisect` is run in the future to find culprits).

**Comment by ikedam** [comment]
I don't think we have to keeping the history clean (it always can be dirty), but I want clarify what changes are actually for this pull request. Actually, we need #36 to have tests pass.
I want to go in this way:

* Reset jenkinsci:feature/35_RecommendedConfigurations to #35 and open merge request with jenkinsci:feature/35_RecommendedConfigurations (A).
* Let you rebase this pull request and update this request.
* Merge this request and merge A.

Let me have some more time to review this request as I'm not sure what happen after resetting feature/35_RecommendedConfigurations (I'm afraid it results this broken and I cannot review changes).

File: pom.xml [hunk 1]
```diff
@@ -3,45 +3,54 @@
...
     </license>
   </licenses>
-  
+
   <properties>
-    <jenkins.version>1.625</jenkins.version><!-- for JENKINS-28298 -->
-    <java.level>7</java.level>
-    <workflow.version>1.0</workflow.version> <!-- only for testing purpose -->
+    <jenkins.version>2.60.3</jenkins.version>
+    <java.level>8</java.level>
```
Comment (ikedam) [inline pom.xml]: I believe you don't need this change for this request.
It would be a part of changes for the upcoming JCaC change.
Would you revert this change to clarify what is actually required to have tests run against latest cores?

Comment (ikedam) [reply pom.xml]: And please keep the comment "only for testing purpose".

File: src/test/java/org/jenkinsci/plugins/authorizeproject/ProjectQueueItemAuthenticatorTest.java [hunk 2]
```diff
@@ -259,6 +258,10 @@ public void testDisabledAtRuntime() throws Exception {
      * Test no exception even if no global-security.jelly is not provided.
      */
     public static class AuthorizeProjectStrategyWithoutGlobalSecurityConfiguration extends AuthorizeProjectStrategy {
+        @DataBoundConstructor
+        public AuthorizeProjectStrategyWithoutGlobalSecurityConfiguration() {
+        }
+
```
Comment (ikedam) [inline src/test/java/org/jenkinsci/plugins/authorizeproject/ProjectQueueItemAuthenticatorTest.java]: (Actually, I'm not sure why tests have passed with older cores....)

File: src/test/java/org/jenkinsci/plugins/authorizeproject/testutil/AuthorizationCheckBuilder.java [hunk 3]
```diff
@@ -41,7 +41,7 @@
  *
  */
 public class AuthorizationCheckBuilder extends Builder {
-    public Authentication authentication = null;
+    public transient Authentication authentication = null;
```
Comment (ikedam) [inline src/test/java/org/jenkinsci/plugins/authorizeproject/testutil/AuthorizationCheckBuilder.java]: Please add a comment that this `transient` is to avoid https://jenkins.io/redirect/class-filter/
as it can be confusing as it looks useless for developers who don't know JEP-200.

File: src/test/java/org/jenkinsci/plugins/authorizeproject/strategy/SpecificUsersAuthorizationStrategyTest.java [hunk 4]
```diff
@@ -864,6 +863,9 @@ public void testConfigurePassword() throws Exception {
             assertEquals("test2", ((SpecificUsersAuthorizationStrategy)p.getProperty(AuthorizeProjectProperty.class).getStrategy()).getUserid());
         }
         
+        // TODO: If JENKINS-59107 is fixed, remove this relogin.
```
Comment (ikedam) [inline src/test/java/org/jenkinsci/plugins/authorizeproject/strategy/SpecificUsersAuthorizationStrategyTest.java]: +1

File: src/test/java/org/jenkinsci/plugins/authorizeproject/strategy/SpecificUsersAuthorizationStrategyTest.java [hunk 5]
```diff
@@ -187,20 +188,18 @@ public void testAuthenticateWithPassword() throws Exception {
         assertTrue(SpecificUsersAuthorizationStrategy.authenticate("test1", false, null, "test1"));
         assertFalse(SpecificUsersAuthorizationStrategy.authenticate("test1", false, null, "test2"));
         assertFalse(SpecificUsersAuthorizationStrategy.authenticate("test1", false, null, ""));
-        assertFalse(SpecificUsersAuthorizationStrategy.authenticate("", false, null, "test2"));
```
Comment (ikedam) [inline src/test/java/org/jenkinsci/plugins/authorizeproject/strategy/SpecificUsersAuthorizationStrategyTest.java]: (Just a note for me)
> The synthetic test case with userId="" has to be removed because the
underlying library does not support this anymore.
```
[ERROR] testAuthenticateWithApitoken(org.jenkinsci.plugins.authorizeproject.strategy.SpecificUsersAuthorizationStrategyTest)  Time elapsed: 3.188 s  <<< ERROR!
com.google.common.util.concurrent.UncheckedExecutionException: java.lang.IllegalArgumentException: Cannot pass null or empty values to constructor
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2234)
        at com.google.common.cache.LocalCache.get(LocalCache.java:3965)
        at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4764)
        at jenkins.security.UserDetailsCache.loadUserByUsername(UserDetailsCache.java:127)
        at hudson.model.User$UserIDCanonicalIdResolver.resolveCanonicalId(User.java:1243)
        at hudson.model.User$CanonicalIdResolver.resolve(User.java:1184)
        at hudson.model.User.get(User.java:509)
        at org.jenkinsci.plugins.authorizeproject.strategy.SpecificUsersAuthorizationStrategy.authenticate(SpecificUsersAuthorizationStrategy.java:125)
        at org.jenkinsci.plugins.authorizeproject.strategy.SpecificUsersAuthorizationStrategyTest.testAuthenticateWithApitoken(SpecificUsersAuthorizationStrategyTest.java:204)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:600)
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
        at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalArgumentException: Cannot pass null or empty values to constructor
        at org.acegisecurity.userdetails.User.<init>(User.java:127)
        at org.jvnet.hudson.test.JenkinsRule$DummySecurityRealm.loadUserByUsername(JenkinsRule.java:933)
        at jenkins.security.UserDetailsCache$Retriever.call(UserDetailsCache.java:172)
        at jenkins.security.UserDetailsCache$Retriever.call(UserDetailsCache.java:161)
        at com.google.common.cache.LocalCache$LocalManualCache$1.load(LocalCache.java:4767)
        at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3568)
        at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2350)
        at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2313)
        at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2228)
        ... 21 more
```

File: src/test/java/org/jenkinsci/plugins/authorizeproject/testutil/AuthorizationCheckBuilder.java (hunk 3 above)
Comment (darxriggs) [reply src/test/java/org/jenkinsci/plugins/authorizeproject/testutil/AuthorizationCheckBuilder.java]: You prefer such information as code comment rather than as commit comment?
I'll add it to the code too.

File: pom.xml (hunk 1 above)
Comment (darxriggs) [reply pom.xml]: I reverted the removal of the "only for testing purpose" comment.

Comment (darxriggs) [reply pom.xml]: I see your point in just wanting to have the minimum required changes. But not performing the Java and Jenkins version change from commit f65754e7c9952adce3aab5ddded5f6ef726e35a1 "Modernize plugin (upgrade parent pom & Java 8 compatibility)" means ripping it apart.

That's something I don't want to do. It would require more effort now and complicates things.

For example:
- The parent "plugin-pom" dropped support for Jenkins 1.x in version 3.x (see https://github.com/jenkinsci/plugin-pom/blob/master/CHANGELOG.md#30).
- Not upgrading Jenkins also means additional entries for `<dependencyManagement>` to satisfy the maven-enforcer-plugin. They should then later be reverted when the upgrade is performed.

I still suggest to first merge #37 standalone and perform all other changes afterwards.
This pull request is currently setup in a way that it exactly contains all commits of #37 (commit 1-4) and additionally adds commits to fix tests for newer Jenkins versions (commit 5-7).

BTW I performed changes similar to #37 for many other plugins over the last year.

**Comment by darxriggs** [comment]
This pull request now contains all changes as I would want them to be merged.

All but one of your review comments have been incorporated. If you insist on not having Java & Jenkins upgraded here I would do it but reluctantly - see https://github.com/jenkinsci/authorize-project-plugin/pull/40#discussion_r319754552.

File: src/test/java/org/jenkinsci/plugins/authorizeproject/testutil/AuthorizationCheckBuilder.java (hunk 3 above)
Comment (ikedam) [reply src/test/java/org/jenkinsci/plugins/authorizeproject/testutil/AuthorizationCheckBuilder.java]: Yes, I prefer that even though I know you described it in the commit comments.
Thank you.

File: pom.xml (hunk 1 above)
Comment (ikedam) [reply pom.xml]: I got your point. Let me have some more time to think over that.

Comment (ikedam) [reply pom.xml]: This is consistent with https://github.com/jenkinsci/plugin-pom/pull/83/files#r146937116 .
I accept it.
